satellite/audit/reservoirservice.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+)
+
+// ReservoirService is a temp name for the service struct during the audit 2.0 refactor.
+// Once V3-2363 and V3-2364 are implemented, ReservoirService will replace the existing Service struct.
+type ReservoirService struct {
+	log *zap.Logger
+
+	nodesToSelect  int
+	PathsToAudit   []storj.Path
+	reservoirSlots int
+	Reservoirs     map[storj.NodeID]*Reservoir
+	rand           *rand.Rand
+
+	MetainfoLoop *metainfo.Loop
+	Loop         sync2.Cycle
+}
+
+// NewReservoirService instantiates Service2
+func NewReservoirService(log *zap.Logger, metaLoop *metainfo.Loop, config Config) (*ReservoirService, error) {
+	return &ReservoirService{
+		log: log,
+
+		nodesToSelect:  config.NodeSelectCount,
+		reservoirSlots: config.Slots,
+		rand:           rand.New(rand.NewSource(time.Now().Unix())),
+
+		MetainfoLoop: metaLoop,
+		Loop:         *sync2.NewCycle(config.Interval),
+	}, nil
+}
+
+// Run runs auditing service 2.0
+func (service *ReservoirService) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("audit 2.0 is starting up")
+
+	return service.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+		pathCollector := NewPathCollector(service.reservoirSlots, service.rand)
+		err = service.MetainfoLoop.Join(ctx, pathCollector)
+		if err != nil {
+			service.log.Error("error joining metainfoloop", zap.Error(err))
+			return nil
+		}
+
+		_, err = service.Select(ctx, pathCollector.Reservoirs)
+		if err != nil {
+			service.log.Error("error selecting reservoirs", zap.Error(err))
+			return nil
+		}
+		return nil
+	})
+}
+
+// Select randomly selects segments to audit
+func (service *ReservoirService) Select(ctx context.Context, reservoirs map[storj.NodeID]*Reservoir) (pathsToAudit []storj.Path, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if len(reservoirs) != 0 {
```

Comment:
Using `for nodeID, reservoir := range reservoirs {` would roughly achieve the same goal, however that's somewhat implementation specific.

The other would be to put all of them into a `[]*Reservoir` and then shuffle the list. No need for the random selection one-by-one.

Reply:
rand.Shuffle should be fine

Reply:
I mean, the point of having multiple paths is to audit all of them, otherwise wouldn't we just select one path to begin with. My idea was to shuffle the reservoirs themselves, not the paths.